advisor-backend/routes/schedule.js

```diff
@@ -39,12 +39,78 @@ const calculateInDegrees = (
   }
 };
 
+const checkDesiredYearAndBefore = (
+  scheduleObject,
+  desiredYear,
+  postReqsSet
+) => {
+  let validMoveFlag = true;
+  for (const year in scheduleObject) {
+    // only check desiredYear and before
+    if (parseInt(year) <= desiredYear) {
+      const coursesSet = scheduleObject[year];
+      postReqsSet.forEach((postReq) => {
+        if (coursesSet.has(postReq)) {
+          // invalid schedule
+          validMoveFlag = false;
+        }
+      });
+    }
+  }
+  return validMoveFlag;
+};
+
+const checkDesiredYearAndAfter = (scheduleObject, desiredYear, preReqsSet) => {
+  let validMoveFlag = true;
+  for (const year in scheduleObject) {
+    // only check desiredYear and after
+    if (parseInt(year) >= desiredYear) {
+      const coursesSet = scheduleObject[year];
+      preReqsSet.forEach((preReq) => {
+        if (coursesSet.has(preReq)) {
+          // invalid schedule
+          validMoveFlag = false;
+        }
+      });
+    }
+  }
+  return validMoveFlag;
+};
+
+const createScheduleObject = (schedule, scheduleObject) => {
+  for (let i = 0; i < schedule.length; i++) {
+    const year = schedule[i];
+    const yearSemester = year.semesters[0];
+    const yearClasses = yearSemester.classes;
+    const yearSet = new Set();
+    for (let j = 0; j < yearClasses.length; j++) {
+      const course = yearClasses[j];
+      yearSet.add(course.id);
+    }
+    scheduleObject[year.number] = yearSet;
+  }
+};
+
```

Comment:
This is fine, but it'd be more "javascripty" if you used [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce) and/or `.map` for this.